---
 gdb/valarith.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

Index: gdb-6.7/gdb/valarith.c
===================================================================
--- gdb-6.7.orig/gdb/valarith.c	2007-10-16 10:37:36.000000000 -0400
+++ gdb-6.7/gdb/valarith.c	2007-10-16 10:38:40.000000000 -0400
@@ -935,7 +935,10 @@ value_binop (struct value *arg1, struct 
 	      break;
 
 	    case BINOP_DIV:
-	      v = v1 / v2;
+	      if (v2 != 0)
+		v = v1 / v2;
+	      else
+		error (_("Division by zero"));
 	      break;
 
 	    case BINOP_EXP:
@@ -946,7 +949,10 @@ value_binop (struct value *arg1, struct 
 	      break;
 
 	    case BINOP_REM:
-	      v = v1 % v2;
+	      if (v2 != 0)
+		v = v1 % v2;
+	      else
+		error (_("Division by zero"));
 	      break;
 
 	    case BINOP_MOD:
