---
 gdb/remote.c |    2 +-
 gdb/symtab.c |    4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

Index: gdb-6.7.1/gdb/remote.c
===================================================================
--- gdb-6.7.1.orig/gdb/remote.c	2007-12-31 13:23:10.000000000 -0500
+++ gdb-6.7.1/gdb/remote.c	2007-12-31 13:27:07.000000000 -0500
@@ -1343,7 +1343,7 @@ unpack_varlen_hex (char *buff,	/* packet
 static char *
 unpack_nibble (char *buf, int *val)
 {
-  ishex (*buf++, val);
+  *val = fromhex (*buf++);
   return buf;
 }
 
Index: gdb-6.7.1/gdb/symtab.c
===================================================================
--- gdb-6.7.1.orig/gdb/symtab.c	2007-12-31 13:23:10.000000000 -0500
+++ gdb-6.7.1/gdb/symtab.c	2007-12-31 13:29:57.000000000 -0500
@@ -2408,6 +2408,8 @@ find_line_common (struct linetable *l, i
   int best_index = -1;
   int best = 0;
 
+  *exact_match = 0;
+
   if (lineno <= 0)
     return -1;
   if (l == 0)
@@ -2433,8 +2435,6 @@ find_line_common (struct linetable *l, i
     }
 
   /* If we got here, we didn't get an exact match.  */
-
-  *exact_match = 0;
   return best_index;
 }
 
