Index: gdb-6.1/gdb/testsuite/gdb.java/jmisc.exp
===================================================================
--- gdb-6.1.orig/gdb/testsuite/gdb.java/jmisc.exp	2004-02-24 11:09:48.000000000 -0500
+++ gdb-6.1/gdb/testsuite/gdb.java/jmisc.exp	2004-04-19 11:00:22.000000000 -0400
@@ -72,7 +72,7 @@ if ![set_lang_java] then {
 
     send_gdb "ptype jmisc\n"   
     gdb_expect {   
-	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+void <init>\\(void\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
+	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+jmisc\\(\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
 	-re ".*$gdb_prompt $"             { fail "ptype jmisc" }
 	timeout { fail "ptype jmisc (timeout)" ; return }
     }
Index: gdb-6.1/gdb/testsuite/gdb.java/jmisc1.exp
===================================================================
--- gdb-6.1.orig/gdb/testsuite/gdb.java/jmisc1.exp	2003-08-14 15:11:10.000000000 -0400
+++ gdb-6.1/gdb/testsuite/gdb.java/jmisc1.exp	2004-04-19 11:00:23.000000000 -0400
@@ -70,7 +70,7 @@ if ![set_lang_java] then {
 
     send_gdb "ptype jmisc\n"   
     gdb_expect {   
-	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+void <init>\\(void\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
+	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+jmisc\\(\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
 	-re ".*$gdb_prompt $"             { fail "ptype jmisc" }
 	timeout { fail "ptype jmisc (timeout)" ; return }
     }
Index: gdb-6.1/gdb/testsuite/gdb.java/jmisc2.exp
===================================================================
--- gdb-6.1.orig/gdb/testsuite/gdb.java/jmisc2.exp	2002-04-29 17:33:03.000000000 -0400
+++ gdb-6.1/gdb/testsuite/gdb.java/jmisc2.exp	2004-04-19 11:00:23.000000000 -0400
@@ -68,7 +68,7 @@ gdb_test "set print sevenbit-strings" ".
 if ![set_lang_java] then {
     send_gdb "ptype jmisc\n"   
     gdb_expect {   
-	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+void <init>\\(void\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
+	-re "type = class jmisc  extends java.lang.Object \{\[\r\n\ \t]+void main\\(java\.lang\.String\\\[]\\);\[\r\n\ \t]+jmisc\\(\\);\[\r\n\ \t]+\}\[\r\n\ \t]+$gdb_prompt $"               { pass "ptype jmisc" }
 	-re ".*$gdb_prompt $"             { fail "ptype jmisc" }
 	timeout { fail "ptype jmisc (timeout)" ; return }
     }
