Update staticthreads.exp to handle debugging info in libpthread.a.

Index: gdb-6.3.90.20051119/gdb/testsuite/gdb.threads/staticthreads.exp
===================================================================
--- gdb-6.3.90.20051119.orig/gdb/testsuite/gdb.threads/staticthreads.exp	2005-11-19 22:56:59.000000000 -0500
+++ gdb-6.3.90.20051119/gdb/testsuite/gdb.threads/staticthreads.exp	2005-11-20 00:01:11.000000000 -0500
@@ -53,6 +53,10 @@ gdb_test_multiple "continue" "$test" {
     -re " sem_post .*$gdb_prompt " {
 	pass "$test"
     }
+    -re " (.*_)sem_post .*$gdb_prompt " {
+	# Glibc uses aliases for internal symbols; match __new_sem_post.
+	pass "$test"
+    }
     -re "Program received signal .*$gdb_prompt " {
 	kfail gdb/1328 "$test"
     }
